modules/lib/lib-portal/src/main/resources/lib/xp/portal.ts

```diff
@@ -83,15 +83,70 @@ export interface SiteConfig<Config> {
     config: Config;
 }
 
-export interface AssetUrlParams {
+export interface Cookie {
+    value: string;
+    path?: string;
+    domain?: string;
+    maxAge?: number;
+    secure?: boolean;
+    httpOnly?: boolean;
+    sameSite?: 'lax' | 'strict' | 'none' | '';
+}
+
+type LiteralUnion<T extends U, U = string> = T | (U & Record<never, never>);
+
+export type RequestParams = Record<string, string | string[] | undefined>;
+
+export interface Request<Params extends RequestParams = RequestParams,
+    Headers extends Record<string, string | undefined> = Record<string, string | undefined>,
+    Cookies extends Record<string, string | Cookie | undefined> = Record<string, string | Cookie | undefined>,
+    > {
+    method: LiteralUnion<'GET' | 'PUT' | 'POST' | 'DELETE' | 'HEAD' | 'PATCH' | 'OPTIONS' | 'TRACE' | 'CONNECT'>;
+    scheme: 'http' | 'https';
+    host: string;
+    port: number;
+    path: string;
+    rawPath: string;
+    url: string;
+    remoteAddress: string;
+    mode: 'inline' | 'edit' | 'preview' | 'live' | 'admin';
+    webSocket: boolean;
+    repositoryId: string;
+    branch: LiteralUnion<'draft' | 'master'>;
+    contextPath: string;
+    body?: string;
+    params: Params;
+    headers: Headers;
+    cookies: Cookies;
+    contentType?: string;
+}
+
+export interface Response<ResponseBody = unknown> {
```

Comment:
Every field in `Response` should be optional, right?

Reply:
We have a dilemma here. Response that controllers receive has many non-optional fields. But the one that controllers could return may have all fields optional (because it is sort-of patch to existing one)

Reply:
It's probably necessary to have two `Response` types. Maybe something like this:

```typescript
export interface InputResponse { // sorry about the bad naming here
  status: number;
  ...
}

export type Response = Partial<InputResponse>;
```

Be careful about naming it `Response` or `Request`, I think there will normally be a naming collision against the global fields in the web platform. (That was why I namespaced it in *enonic-types*)

Reply:
> naming collision against the global fields in the web platform
Do you mean tsconfig lib: ['DOM']?

I think using request is fine, because we are talking about Enonic XP server-side files.

It's a bit tricky to "remove" the DOM from default environment.

I found that adding this:
node_modules/typescript/lib/lib.es5.d.ts

To the includes files does the trick.

It should probably be used in combination with nolib: true